python/dgl/transform.py

```diff
@@ -2482,4 +2484,111 @@ def as_immutable_graph(hg):
                 '\tdgl.as_immutable_graph will do nothing and can be removed safely in all cases.')
     return hg
 
+def sort_out_edges(g, tag=None, tag_pos="_TAG_POS"):
+    """A copy of the given graph whose out edges are sorted.
+    After sorting, edges whose destination shares the same
+    tag will be arranged in a consecutive range. Note that this
+    will not change the edge ID. It only changes the order in the
+    internal CSR storage. As such, the graph must allow CSR storage.
```

Comment:
Suggested change

```python
def sort_out_edges(g, tag, tag_offset_name='_TAG_OFFSET'):

"""Return a new graph which sorts the out edges of each node.

Sort the out edges according to the given destination node tags in integer.
A typical use case is to sort the edges by the destination node types, where
the tags represent destination node types. After sorting, edges sharing
the same tag will be arranged in a consecutive range in
a node's adjacency list. Following is an example:

    Consider a graph as follows:

    0 -> 0, 1, 2, 3, 4
    1 -> 0, 1, 2

    Given node tags [1, 1, 0, 2, 0], each node's adjacency list
    will be sorted as follows:

    0 -> 2, 4, 0, 1, 3
    1 -> 2, 0, 1

The function will also returns the starting offsets of the tag
segments in a tensor of shape `(N, max_tag+2)`. For node `i`,
its out-edges connecting to node tag `j` is stored between
`tag_offsets[i][j]` ~ `tag_offsets[i][j+1]`. Since the offsets
can be viewed node data, we store it in the
`ndata` of the returned graph. Users can specify the
ndata name by the `tag_pos_name` argument.

Note that the function will not change the edge ID neither
how the edge features are stored. The input graph must
allow CSR format. Graph must be on CPU.

If the input graph is heterogenous, it must have only one edge
type and two node types (i.e., source and destination node types).
In this case, the provided node tags are for the destination nodes,
and the tag offsets are stored in the source node data.

The sorted graph and the calculated tag offsets are needed by
certain operators that consider node tags. See xxx for an example.

Examples
-----------
...

Parameters
------------
g : DGLGraph
    The input graph.
tag : Tensor
    Integer tensor of shape `(N,)`, `N` being the number of (destination) nodes.
tag_offset_name : str
    The name of the node feature to store tag offsets.

Returns
-------
g_sorted : DGLGraph
    A new graph whose out edges are sorted. The node/edge features of the
    input graph is shallow-copied over.
    - `g_sorted.ndata[tag_offset_name]` : Tensor of shape `(N, max_tag + 2)`. If
       `g` is heterogeneous, get from `g_sorted.srcdata`.
"""
```

Reply:
No need to support None tag at the moment. We could add it later when there is a request.